[KEP-4017]: update e2e and integration test for PodIndexLabel

alaypatel07 opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->
/kind cleanup

#### What this PR does / why we need it:
Required for driving [KEP-4017](https://github.com/kubernetes/enhancements/issues/4017) to GA


#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".
NONE

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs
- [KEP]: https://github.com/kubernetes/enhancements/blob/master/keps/sig-apps/4017-pod-index-label/README.md
```

alaypatel07 committed:
update e2e and integration test for PodIndexLabel

Signed-off-by: Alay Patel <alayp@nvidia.com>

kannon92 commented on test/integration/job/job_test.go:
```diff
@@ -4127,13 +4127,20 @@ func validateIndexedJobPods(ctx context.Context, t *testing.T, clientSet clients
 		if metav1.IsControlledBy(&pod, jobObj) {
 			if pod.Status.Phase == v1.PodPending || pod.Status.Phase == v1.PodRunning {
-				ix, err := getCompletionIndex(&pod)
+				annotationIx, err := getCompletionIndex(pod.Annotations)
```
The KEP is PodIndexLabel.. I think we should probably grab the labels of a pod rather than the annotations.

kannon92 replied:
Sorry I see below that you grabbing the label.

kannon92 commented:
/ok-to-test

alaypatel07 commented:
This failure https://prow.k8s.io/view/gs/kubernetes-ci-logs/pr-logs/pull/128107/pull-kubernetes-integration/1848357154730283008 was programmer error, I used wrong label key in the test. Fixed it, triggering the test again.

/test pull-kubernetes-integration

kannon92 commented:
LGTM to me!

I think you will need a sig-apps person to sign off on this.

/cc @atiratree @mimowo

alaypatel07 commented:
```
2024-10-21T20:23:37.160779432Z stderr F I1021 20:23:37.159790       1 flags.go:64] FLAG: --feature-gates=":AllAlpha=false,:AllBeta=false"
```

The [job](https://prow.k8s.io/view/gs/kubernetes-ci-logs/pr-logs/pull/128107/pull-kubernetes-conformance-kind-ga-only-parallel/1848456471759556608) failed but I see controller manager logs disabling AllBeta features https://storage.googleapis.com/kubernetes-ci-logs/pr-logs/pull/128107/pull-kubernetes-conformance-kind-ga-only-parallel/1848456471759556608/artifacts/kind-control-plane/pods/kube-system_kube-controller-manager-kind-control-plane_046aec42e7ddb18217c94af6f129b41e/kube-controller-manager/0.log.20241021-202712

@mimowo or @soltysh can you please look if I am missing anything above?

kannon92 commented:
I'm not entirely sure what sig-apps does for features that are supposed to go GA. I dont' know if this should be made conformance.

You could try opening up a separate PR with feature gate promotion to stable and add these e2e tests to see if this is related to the test-infra feature gate turning this feature off.

alaypatel07 commented:
@kannon92 Verified locally with the same feature flag configuration as observed in the test logs.
```
$ k logs -f -n kube-system kube-controller-manager-kind-1.31-dra-control-plane  | grep feature
I1021 21:20:03.352568       1 flags.go:64] FLAG: --feature-gates=":AllAlpha=false,:AllBeta=false"

$ k get pods --show-labels
NAME                          READY   STATUS      RESTARTS   AGE   LABELS
busybox-indexed-job-0-56fx4   0/1     Completed   0          5s    batch.kubernetes.io/controller-uid=1c0101cb-3ba7-42d6-9900-f6e1168fbbce,batch.kubernetes.io/job-name=busybox-indexed-job,controller-uid=1c0101cb-3ba7-42d6-9900-f6e1168fbbce,job-name=busybox-indexed-job
busybox-indexed-job-1-vl6tj   0/1     Completed   0          5s    batch.kubernetes.io/controller-uid=1c0101cb-3ba7-42d6-9900-f6e1168fbbce,batch.kubernetes.io/job-name=busybox-indexed-job,controller-uid=1c0101cb-3ba7-42d6-9900-f6e1168fbbce,job-name=busybox-indexed-job
busybox-indexed-job-2-xnnsx   0/1     Completed   0          5s    batch.kubernetes.io/controller-uid=1c0101cb-3ba7-42d6-9900-f6e1168fbbce,batch.kubernetes.io/job-name=busybox-indexed-job,controller-uid=1c0101cb-3ba7-42d6-9900-f6e1168fbbce,job-name=busybox-indexed-job
```

The feature gate is off and hence we dont see those labels.  Trying to look for how to turn the feature gate on for the test.

alaypatel07 commented:
I think the test needs to be executed with this setup: https://github.com/kubernetes/test-infra/blob/d660515b07f96065115fc17e5e5c9ed2a5df0263/config/jobs/kubernetes/sig-testing/kubernetes-kind.yaml#L334

/test pull-kubernetes-e2e-kind-beta-features

alaypatel07 commented:
https://prow.k8s.io/view/gs/kubernetes-ci-logs/pr-logs/pull/128107/pull-kubernetes-e2e-kind-beta-features/1848487512914268160

This passed so we are not the right track. I think I need to figure out how to not invoke the new e2e test with ga conformance test environment

mimowo commented:
> This passed so we are on the right track. I think I need to figure out how to not invoke the new e2e test with ga conformance test environment

The suite `pull-kubernetes-conformance-kind-ga-only-parallel` only enables GA features, no way to enable some of them.

Also, you cannot create a new test straight as conformance. The steps should be:
- create a regular e2e test, best if ready to be promoted without code changes inside
- promote the feature-gate to GA
- promote the test (at least 2 weeks after the test was created)

More details here: https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md. In particular in the requirements section you will find "it is stable and runs consistently (e.g., no flakes), and has been running for at least two weeks".

mimowo commented on test/e2e/apps/job.go:
```diff
@@ -466,6 +467,44 @@ done`}
+			Description: Create an Indexed job. Job MUST complete successfully.
+			Ensure that created pods have completion index label.
+	*/
+	framework.ConformanceIt("should create pods for an Indexed job with completion indexes for PodLabelIndex feature", feature.PodIndexLabel, func(ctx context.Context) {
```
```suggestion
	gomega.It("should create pods with completion indexes for an indexed Job", feature.PodIndexLabel, func(ctx context.Context) {
```

alaypatel07 replied:
it looks like `framework.It` is available as wrapper to gomega.It, used that instead.

mimowo commented on test/e2e/apps/job.go:
```diff
@@ -466,6 +467,44 @@ done`}
+
+		ginkgo.By("Creating Indexed job")
+		job := e2ejob.NewTestJob("succeed", "indexed-job", v1.RestartPolicyNever, parallelism, completions, nil, backoffLimit)
+		mode := batchv1.IndexedCompletion
+		job.Spec.CompletionMode = &mode
```
```suggestion
		job.Spec.CompletionMode = ptr.To(batchv1.IndexedCompletion)
```

alaypatel07 replied:
updated

mimowo commented on test/e2e/apps/job.go:
```diff
@@ -466,6 +467,44 @@ done`}
+		framework.ExpectNoError(err, "failed to get pod list for job in namespace: %s", f.Namespace.Name)
+		succeededIndexes := sets.NewInt()
+		for _, pod := range pods.Items {
+			if pod.Status.Phase == v1.PodSucceeded && pod.Labels != nil {
```
I don't think we need the `pod.Status.Phase == v1.PodSucceeded` check. I know that technically a pod still can be seen as Running due to how event propagation works (maybe Job controller already received the Succeded event, but not the test code), but even if it is running it already has the index, so let's check it.

Also, `pod.Labels != nil` seems extra.

alaypatel07 replied:
+1, removed the pod status and pod label checks. The test is already asserting job is complete, indices will be available

mimowo commented on test/e2e/apps/job.go:
```diff
@@ -466,6 +467,44 @@ done`}
+		for _, pod := range pods.Items {
+			if pod.Status.Phase == v1.PodSucceeded && pod.Labels != nil {
+				ix, err := strconv.Atoi(pod.Labels[batchv1.JobCompletionIndexAnnotation])
+				framework.ExpectNoError(err, "failed obtaining completion index with label key: %s from pod in namespace: %s", batchv1.JobCompletionIndexAnnotation, f.Namespace.Name)
```
```suggestion
				framework.ExpectNoError(err, "failed obtaining completion index for Pod: %s", klog.KObj(pod))
```
I don't think the name of the label helps here

alaypatel07 replied:
I saw klog was not imported, so I changed it following the pattern used in the rest of the test code
```
framework.ExpectNoError(err, "failed obtaining completion index in namespace: %s for Pod: %s", pod.Namespace, pod.Name)
```

PTAL

mimowo commented on test/e2e/apps/job.go:
```diff
@@ -466,6 +467,44 @@ done`}
+		}
+		gotIndexes := succeededIndexes.List()
+		wantIndexes := []int{0, 1, 2, 3}
+		gomega.Expect(gotIndexes).To(gomega.Equal(wantIndexes), "expected completed indexes for labely key %s: %s, but got %s", batchv1.JobCompletionIndexAnnotation, wantIndexes, gotIndexes)
```
```suggestion
		gomega.Expect(gotIndexes).To(gomega.Equal(wantIndexes), "expected completed indexes for Job: %s", klog.KObj(job))
```
IIUC the want and got would be already output by gomega (but please confirm). Getting to know the job, or at least the namespace is important too.

alaypatel07 replied:
+1 it is already output by gomega

mimowo commented on test/e2e/apps/statefulset.go:
```diff
@@ -1156,6 +1157,45 @@ var _ = SIGDescribe("StatefulSet", func() {
+			}
+			wantIndexes := []int{0, 1, 2}
+			gotIndexes := labelIndices.List()
+			gomega.Expect(gotIndexes).To(gomega.Equal(wantIndexes), "expected indexes for label with key %s: %s, but got %s", appsv1.PodIndexLabel, wantIndexes, gotIndexes)
```
as for Job, please check if the want and got is already output by gomega, if it is, then not need to repeat. Instead output the sts name and namespace.

alaypatel07 replied:
updated

mimowo commented on test/e2e/apps/statefulset.go:
```diff
@@ -1156,6 +1157,45 @@ var _ = SIGDescribe("StatefulSet", func() {
+			Description: When Pods for a StatefulSet is created it MUST be created with PodIndexLabel feature
+			Attempt to read the PodIndexLabel key and value
+		*/
+		framework.ConformanceIt("pods for Statefulset should have PodIndexLabel with right index", feature.PodIndexLabel, func(ctx context.Context) {
```
as for Job, don't make it conformance yet

mimowo commented on test/e2e/apps/statefulset.go:
```diff
@@ -1156,6 +1157,45 @@ var _ = SIGDescribe("StatefulSet", func() {
+			labelIndices := sets.NewInt()
+			for _, pod := range pods.Items {
+				ix, err := strconv.Atoi(pod.Labels[appsv1.PodIndexLabel])
+				framework.ExpectNoError(err, "failed obtaining label with key: %s from pod: %s in namespace: %s", appsv1.PodIndexLabel, pod.Name, pod.Namespace)
```
```suggestion
				framework.ExpectNoError(err, "failed obtaining pod index for pod: %s", klog.KObj(pod))
```

mimowo commented on test/e2e/apps/job.go:
```diff
@@ -466,6 +467,44 @@ done`}
+		err = e2ejob.WaitForJobComplete(ctx, f.ClientSet, f.Namespace.Name, job.Name, nil, completions)
+		framework.ExpectNoError(err, "failed to ensure job completion in namespace: %s", f.Namespace.Name)
+
+		ginkgo.By("Ensuring pods with index for job exist")
```
```suggestion
		ginkgo.By("Ensuring all pods have the required index labels")
```

mimowo commented on test/integration/statefulset/statefulset_test.go:
```diff
@@ -655,42 +656,47 @@ func TestDeletingPodForRollingUpdatePartition(t *testing.T) {
+		namespace           string
+		replicas            int
+		expectedPodOrdinals []string
+		podNamePrefix       string
```
It does not seem needed as all statefulstes have the same prefix in this test, if so then drop

mimowo commented on test/integration/statefulset/statefulset_test.go:
```diff
@@ -655,42 +656,47 @@ func TestDeletingPodForRollingUpdatePartition(t *testing.T) {
+		name                string
+		namespace           string
+		replicas            int
+		expectedPodOrdinals []string
```
```suggestion
		expectedPodIndexes []string
```
nit: to better correlate the naming with the feature name, and the label name

mimowo commented on test/integration/statefulset/statefulset_test.go:
```diff
@@ -719,17 +725,43 @@ func TestStatefulSetStartOrdinal(t *testing.T) {
 			ignoreOrder := cmpopts.SortSlices(func(a, b string) bool {
 				return a < b
 			})
+			ignoreOrderForOrdinals := cmpopts.SortSlices(func(a, b string) bool {
```
Hm, this is verbose (repeated conversion for a and b), maybe just read all the indexes as int, and then sort ints.

mimowo reviewed (commented):
initial pass

mimowo commented:
/release-note-none
Adding new tests is not "user facing" change

alaypatel07 commented:
@mimowo addressed all the feedback, I am assuming changing from framework.ConformanceIt to framework.It will make the  failing job`pull-kubernetes-conformance-kind-ga-only-parallel` go green.

Please take another look, TIA

alaypatel07 commented:
```
Kubernetes e2e suite: [It] [sig-node] [Serial] Pod InPlace Resize Container (scheduler-focused) [Feature:InPlacePodVerticalScaling] pod-resize-scheduler-tests 
```

The failed test seems unrelated to the code changes. https://prow.k8s.io/view/gs/kubernetes-ci-logs/pr-logs/pull/128107/pull-kubernetes-e2e-gce-cos-alpha-features/1848781933614469120

soltysh commented on test/e2e/apps/job.go:
```diff
@@ -466,6 +467,41 @@ done`}
+			Description: Create an Indexed job. Job MUST complete successfully.
+			Ensure that created pods have completion index label.
+	*/
+	framework.It("should create pods with completion indexes for an Indexed Job", feature.PodIndexLabel, func(ctx context.Context) {
```
After this test settles, when you consider moving this to conformance, I'd squash this with this one: https://github.com/kubernetes/kubernetes/blob/d9b95ea94f41b942440adbdfdbe35fb78a2ed6f7/test/e2e/apps/job.go#L437-L469
because both are almost identical, the only difference is one is checking annotation and this one labels, which both should match.

soltysh replied:
Can you add a TODO here, so that we don't forget to do that?

soltysh commented on test/e2e/apps/statefulset.go:
```diff
@@ -1156,6 +1157,45 @@ var _ = SIGDescribe("StatefulSet", func() {
+			Description: When Pods for a StatefulSet is created it MUST be created with PodIndexLabel feature
+			Attempt to read the PodIndexLabel key and value
+		*/
+		framework.It("pods for Statefulset should have PodIndexLabel with right index", feature.PodIndexLabel, func(ctx context.Context) {
```
I'm inclined to push the label check into this test: https://github.com/kubernetes/kubernetes/blob/d9b95ea94f41b942440adbdfdbe35fb78a2ed6f7/test/e2e/apps/statefulset.go#L144-L187
which checks basic statefulset functionality.

soltysh reviewed (commented):
A few minor comments, but mostly looks good.

soltysh commented:
/assign

soltysh commented on test/e2e/apps/job.go:
```diff
@@ -466,6 +467,42 @@ done`}
+			Description: Create an Indexed job. Job MUST complete successfully.
+			Ensure that created pods have completion index label.
+	*/
+	// TODO: once this test is stable, promote it to the conformance test named "Ensure Pods of an Indexed Job get a unique index."
```
```suggestion
	// TODO: once this test is stable, squash the functionality into pre-existing conformance test called "should create pods for an Indexed job with completion indexes and specified hostname" earlier in this file. 
```

alaypatel07 replied:
updated

soltysh commented on test/e2e/apps/statefulset.go:
```diff
@@ -1156,6 +1157,46 @@ var _ = SIGDescribe("StatefulSet", func() {
+			e2estatefulset.WaitForRunningAndReady(ctx, c, *ss.Spec.Replicas, ss)
+			waitForStatus(ctx, c, ss)
+
+			ginkgo.By("checking the index label and value of all pods")
```
I've meant to move this checks into that "should provide basic identity" test.

soltysh reviewed (approved):
/lgtm
/approve
